library/aes.c

```diff
@@ -84,8 +84,10 @@ static int aes_padlock_ace = -1;
 /*
  * Forward S-box
  */
-#if !defined(MBEDTLS_AES_ENCRYPT_ALT) || !defined(MBEDTLS_AES_SETKEY_ENC_ALT) || \
-    !defined(MBEDTLS_AES_SETKEY_DEC_ALT)
+#if !defined(MBEDTLS_AES_ENCRYPT_ALT) || \
+    (!defined(MBEDTLS_AES_SETKEY_ENC_ALT) && (!defined(MBEDTLS_AES_USE_HARDWARE_ONLY) || \
+    !defined(MBEDTLS_AES_ROM_TABLES))) || \
+    (!defined(MBEDTLS_AES_SETKEY_DEC_ALT) && !defined(MBEDTLS_AES_USE_HARDWARE_ONLY))
```

Comment:
this is pretty much unreadable

The only way of making it readable I found is to have too long a line, so no good suggestion here.

```
#if !defined(MBEDTLS_AES_ENCRYPT_ALT) || \                                                                                                                                   
    (!defined(MBEDTLS_AES_SETKEY_ENC_ALT) && (!defined(MBEDTLS_AES_USE_HARDWARE_ONLY) || !defined(MBEDTLS_AES_ROM_TABLES))) || \                                             
    (!defined(MBEDTLS_AES_SETKEY_DEC_ALT) && !defined(MBEDTLS_AES_USE_HARDWARE_ONLY))
```

Reply:
I considered simply marking these variables as "maybe unused" with `__attribute__((unused))`, and then all the variable declarations can just be unconditionally present. They are static so there's no impact on code size.

The places where they are used remain awkward though.

I was considering a separate PR for this, but I could bring it into this one if you prefer?

Reply:
I think marking them as "maybe unused" is even worse - the idea of this PR is (to me) to get what's needed correctly specified. Just saying "here's some data, please, compiler, remove it if it's not needed and don't tell me about it" seems... wrong.

I was wondering about building up a `#define NEED_F_SBOX` or some such, but it doesn't really help the readability.

I guess we're stuck with this one

Reply:
I also explored using `NEED_XYZ`, but agreed, it doesn't help very much.

Actually I do think leaving it to the compiler helps - it greatly simplifies the definitions. I don't think it can introduce any risk?

Reply:
I've pushed it anyway; see what you think. On balance I think it is definitely an improvement - why manually maintain complex `#if ...` blocks when the compiler can do a more reliable job with no complexity in the codebase?

Reply:
In general I prefer to let the compiler figure out details like eliminating variables that are not used in some configurations, eliminating checks that are no-ops in some configurations, etc. The problem is to disable the warnings on all the compilers we care about.